src/main/cpp/src/cast_string_to_datetime.cu

```diff
@@ -857,6 +860,169 @@ std::unique_ptr<cudf::column> parse_ts_strings(cudf::strings_column_view const&
+  if ('-' == sign_c || '+' == sign_c) {
+    ++pos;
+    if ('-' == sign_c) { negative_year_sign = true; }
+  }
+
+  // parse year: yyyy[y][y][y]
+  if (!parse_int(ptr,
+                 pos,
+                 end_pos,
+                 d.year,
+                 /*min_digits*/ 4,
+                 /*max_digits*/ 7)) {
+    return false;
+  }
+  if (negative_year_sign) { d.year = -d.year; }
+
+  if (eof(pos, end_pos)) {
+    // only has: yyyy[y][y][y], return early
+    return true;
+  }
+
+  // parse month: -[m]m
+  if (!parse_char(ptr, pos, '-') || !parse_int(ptr,
+                                               pos,
+                                               end_pos,
+                                               d.month,
+                                               /*min_digits*/ 1,
+                                               /*max_digits*/ 2)) {
+    return false;
+  }
+
+  if (eof(pos, end_pos)) {
+    // only has: yyyy[y][y]-[m]m, return early
+    return true;
+  }
+
+  // parse day: -[d]d
+  if (!parse_char(ptr, pos, '-') || !parse_int(ptr,
+                                               pos,
+                                               end_pos,
+                                               d.day,
+                                               /*min_digits*/ 1,
+                                               /*max_digits*/ 2)) {
+    return false;
+  }
+
+  if (eof(pos, end_pos)) {
+    // no tailing characters, e.g.: "2025-01-01"
+    return true;
+  }
+
+  // parse date time separator
+  // ignore the tailing chars after ' ' or 'T', e.g.: the following are valid:
+  // "2025-01-01T", "2025-01-01Txxxx", and "2025-01-01 xxx"
+  return parse_date_time_separator(ptr, pos);
+}
+
+struct parse_string_to_date_fn {
+  // input strings
+  cudf::column_device_view d_strings;
+
+  // output columns
+  uint8_t* null_mask;
+  cudf::timestamp_D* output;
+
+  __device__ void operator()(cudf::size_type const idx) const
+  {
+    auto const str                   = d_strings.element<cudf::string_view>(idx);
+    unsigned char const* str_ptr     = reinterpret_cast<unsigned char const*>(str.data());
+    unsigned char const* str_end_ptr = str_ptr + str.size_bytes();
+
+    // parse the date string to segments
+    // No need to check null for the `str` element
+    // Because get element on null will return empty string and then result in invalid
+    spark_rapids_jni::date_segments date_segments;
+    auto result_success = parse_date(str_ptr, str_end_ptr, date_segments);
+
+    if (!result_success || !date_segments.is_valid_date()) {
+      null_mask[idx] = 0;
+      output[idx]    = cudf::timestamp_D{cudf::duration_D{0}};
+      return;
+    }
+
+    int64_t days = date_segments.to_epoch_day();
+    // Spark stores date int int, so need to check the range
+    if (days < cuda::std::numeric_limits<int32_t>::min() ||
+        days > cuda::std::numeric_limits<int32_t>::max()) {
+      // exceeds int range.
+      null_mask[idx] = 0;
```

Comment:
Here we output validity instead of null mask. The null mask is set (bit `1`) when the row is null. So the variable needs to be renamed to `validity` instead.